runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -409,6 +423,42 @@ function LanguageTree:_add_injections()
   return query_time
 end
 
+--- Run an asynchronous parse, calling {on_parse} when complete.
+---
+--- @param range boolean|Range?
+--- @param on_parse fun(trees?: table<integer, TSTree>, err?: string)
+--- @return boolean async true if parsing asynchronously
+function LanguageTree:_async_parse(range, on_parse)
+  local ct = self:_buf().changedtick
+  local total_parse_time = 0
+  local redrawtime = vim.o.redrawtime
+
+  local function step()
+    -- If buffer was changed in the middle of parsing, reset parse state
+    if self:_buf().changedtick ~= ct then
+      self._parser:reset()
+      self:invalidate()
+      ct = self:_buf().changedtick
+      total_parse_time = 0
+    end
+
+    local parse_time, trees, finished = tcall(self._parse, self, range, default_parse_timeout_ms)
+    total_parse_time = total_parse_time + parse_time
+
+    if finished then
+      on_parse(trees)
+    elseif total_parse_time > redrawtime then
+      on_parse(nil, 'TIMEOUT')
+    else
+      vim.schedule(step)
+      return true
+    end
+    return false
```

Comment:
I think it's better to return the trees if we have them, and otherwise return nil.

This would chamge how tjw highligher interprets the timeput case, but it should be fine either way.